man/man1/page.1
.TH PAGE 1
.SH NAME
page \- view
FAX,
image, graphic, PostScript, PDF, and
typesetter output
files
.SH SYNOPSIS
.B page
[
.B -abirPRvVw
]
[
.B -p
.I ppi
]
[
.IR file ...
]
.SH DESCRIPTION
.I Page
is a general purpose document viewer.
It can be used to display the individual pages
of a
PostScript,
PDF,
or
.MR troff 1
or
Unix's
.IR tex (1)
device-independent output
file.
.I Troff
or
.I tex
output is simply converted to PostScript in order to be viewed.
It can also be used to view any number of
graphics files
(such as a
FAX
page,
a Plan 9
.MR image 7
file, an Inferno bitmap file, or other common format).
.I Page
displays these
in sequence.
In the absence of named files,
.I page
reads one from standard input.
.PP
By default,
.I page
runs in the window in which it is started
and leaves the window unchanged.
The
.B -R
option causes
.I page
to grow the window if necessary
to display the page being viewed.
The
.B -w
option causes
.I page
to create a new window for itself.
The newly created window will grow as under the
.B -R
option.
If being used to display
multipage documents,
only one file may be specified on the command line.
.PP
The
.B -p
option sets the resolution for PostScript and PDF
files, in pixels per inch.
The default is 100 ppi.
The
.B -r
option reverses the order in which pages are displayed.
.PP
When viewing a document,
.I page
will try to guess the true bounding box, usually rounding up from
the file's bounding box to
8½×11 or A4 size.
The
.B -b
option causes it to respect the bounding box given in the file.
As a more general problem,
some PostScript files claim to conform to Adobe's
Document Structuring Conventions but do not.
The
.B -P
option enables a slightly slower and slightly more
skeptical version of the PostScript processing code.
Unfortunately, there are PostScript documents
that can only be viewed with the
.B -P
option, and there are PostScript documents that
can only be viewed without it.
.PP
When viewing images with
.IR page ,
it listens to the
.B image
plumbing channel
(see
.MR plumber 4 )
for more images to display.
The
.B -i
option causes
.I page
to not load any graphics files nor to read
from standard input but rather to listen
for ones to load from the plumbing channel.
.PP
The
.B -v
option turns on extra debugging output, and
the
.B -V
option turns on even more debugging output.
The
.B -a
option causes
.I page
to call
Unix's
.IR abort (3)
rather than exit cleanly on errors,
to facilitate debugging.
.PP
Pressing and holding button 1 permits panning about the page.
.PP
Button 2 raises a menu of operations on the current image or the
entire set.  The image transformations are non-destructive and are
valid only for the currently displayed image.  They are lost as soon
as another image is displayed.
The button 2 menu operations are:
.TF Resize
.TP
.B Orig size
Restores the image to the original. All modifications are lost.
.TP
.B Zoom
Prompts the user to sweep a rectangle on the image which is
expanded proportionally to the rectangle.
.TP
.B Fit window
Resizes the image so that it fits in the current window.
.TP
.B Rotate 90
Rotates the image 90 degrees clockwise
.TP
.B Upside down
Toggles whether images are displayed upside-down.
.TP
.B Next
Displays the next page.
.TP
.B Prev
Displays the previous page.
.TP
.B Zerox
Displays the current image in a new page window.
Useful for selecting important pages from large documents.
.TP
.B Reverse
Reverses the order in which pages are displayed.
.TP
.B Write
Writes the image to file.
.PD
.PP
Button 3 raises a menu of the
pages
to be selected for viewing in any order.
.PP
Typing a
.B q
or
control-D exits the program.
Typing a
.B u
toggles whether images are displayed upside-down.
(This is useful in the common case of mistransmitted upside-down faxes).
Typing a
.B r
reverses the order in which pages are displayed.
Typing a
.B w
will write the currently viewed page to a new file as a compressed
.MR image 7
file.
When possible, the filename is of the form
.IR basename . pagenum . bit .
Typing a
.B d
removes an image from the working set.
.PP
To go to a specific page, one can type its number followed by enter.
Typing left arrow, backspace, or minus displays the previous page.
Typing right arrow, space, or enter displays the next page.
The up and down arrow pan up and down one half screen height,
changing pages when panning off the top or bottom of the page.
.PP
.I Page
calls
Unix's
.IR gs (1)
to draw each page of PostScript
and
PDF
.IR files .
It also calls a variety of conversion programs, such as those described in
.MR jpg 1 ,
to convert the various raster graphics formats
into Inferno bitmap files.
Pages are converted ``on the fly,'' as needed.
.SH EXAMPLES
.TP
.L
page /sys/src/cmd/gs/examples/tiger.eps
Display a color PostScript file.
.TP
.L
page /usr/inferno/icons/*.bit
Browse the Inferno bitmap library.
.TP
.L
man -t page | page -w
Preview this manual in a new window.
.SH "SEE ALSO
.MR gs 1 ,
.MR jpg 1 ,
.MR proof 1 ,
.MR tex 1 ,
.MR troff 1
.SH SOURCE
.B \*9/src/cmd/page
.SH DIAGNOSTICS
The mouse cursor changes to an arrow and ellipsis
when
.I page
is reading or writing a file.
.SH BUGS
.I Page
supports reading of only one document
file at a time, and
the user interface is clumsy when viewing very large documents.
.PP
When viewing multipage PostScript files that do not contain
.RB `` %%Page ''
comments, the button 3 menu only contains
``this page'' and ``next page'':
correctly determining
page boundaries in Postscript code is not computable
in the general case.
.PP
If
.I page
has trouble viewing a Postscript file,
it might not be exactly conforming: try viewing it with the
.B -P
option.
.PP
The interface to the plumber is unsatisfactory.  In particular,
document references cannot be sent
via plumbing messages.
.PP
There are too many keyboard commands and menu items.
.PP
Displaying a PostScript or PDF file depends both on having
GhostScript
(see
.MR gs 1 )
installed and on the underlying operating system
providing a file descriptor device tree at
.BR /dev/fd .
.PP
Some FreeBSD installations
do not provide file descriptors greater than 2
in
.BR /dev/fd .
To fix this, add
.IP
.EX
/fdescfs    /dev/fd    fdescfs    rw    0    0
.EE
.LP
to
.BR /etc/fstab ,
and then
.B mount
.BR /dev/fd .
(Adding the line to
.B fstab
ensures causes FreeBSD to mount the file system
automatically at boot time.)


CHANGES
# Note: the name listed next to a change sometimes
# submitted the patch but more often just reported
# the problem.

# For entries more recent than January 30, 2008,
# see the Git repository history at 
# https://github.com/9fans/plan9port/.

January 30, 2008
	acme: fix regexp match choice bug
	devdraw: add F11 for full-screen toggle
	rio: accomodate full-screen apps more gracefully
	sam: fix regexp match choice bug

January 28, 2008
	xd: fix printing of 4-byte values on 64-bit machines (Sam Hopkins)

January 12, 2008
	delatex: reset lexer state at each beginning of file
	ipso: do not flush factotum keys unless using factotum file
	libdiskfs: fix ext2 directory reading bug (Mike Mammarella)
	lp: move devices to devices.sample
	rio: make full-screen work properly; add showevent
	venti: fewer calls to gettimeofday
	vmount: need to say -t nfs now on Linux

January 10, 2008
	acme: make STACK bigger for OS X (David Swasey)
	dist,unix: use MKSHELL=$PLAN9/bin/rc (John DeGood)
	libdraw: fix spinning event loop (David Bulkow)
	libregexp: fix match choice bug

December 7, 2007
	acme, sam, libregexp: regexp fixes and reversions

December 4, 2007
	hoc: print exact floating point output
	sam: turn off DEBUG logging (oops)
	vbackup: fix uninitialized variable use in -i code

November 28, 2007
	acme: refresh window properly after Tab command (Fazlul Shahriar)

November 27, 2007
	.hgignore: ignore more
	libdiskfs: make ext2 use little-endian disk format on big-endian machines (Joshua Wood)
	acme: accept clicks between body and scroll bar (Axel Belinfante)
	tpic: clean up better
	upas/smtp: remove intermediate files
	vbackup: add -i flag to avoid score prefetching (David Swasey)
	venti: add debugging HTML form

November 25, 2007
	acme: avoid infinite loop on search in 8125-byte file (Roger Peppe)
	lib9: avoid DIOCGMEDIASIZE on FreeBSD 4.11 (Rodrigo Readi)
	mailfs: use root in imapcopylist (Richard Bilson)
	moveplan9.files: add tmac/tmac.an (Tony Lainson)

November 23, 2007
	9pfuse: ignore Linux O_CLOEXEC flag (Michael Kaminsky)

November 22, 2007
	9pfuse: fix array bounds error (Michael Teichgräber)
	vbackup: fix fsysopen error message (Joshua Wood)

November 7, 2007
	9fs: use -n to mount sources (Jeff Sickel),
	9pfuse: MacFUSE moved again! (Jeff Sickel)

November 6, 2007
	.cvsignore: add 9pm (for old trees)
	unvf: ignore case in Content-Transfer-Encoding: line.
	validateattachment: reject malformed zip files

November 5, 2007
	building: use ar s flag to avoid ranlib (OS X)
	diff: continue after i/o errors when diffing many files
	factotum: silence gcc warning
	getns: turn / into _ in $DISPLAY
	libdiskfs: OS X linker fixes
	libmach: OS X linker fixes
	libmp: no text jump labels (OS X)
	libthread: use our own swapcontext on OS X
	mailfs: handle NIL body, add -r flag (Richard Bilson)

November 4, 2007
	man: document hg interface, fix CVS host name

November 3, 2007
	libthread: more OS X 10.5 changes (Jeff Sickel)

October 31, 2007
	venti: use 9ar, not ar (fixes Darwin build)

October 30, 2007
	libthread: OS X 10.5 tweaks (Jeff Sickel, Bakul Shah)

October 29, 2007
	grap: fclose, error context bug fixes (Doug Gwyn)
	venti: fix sync deadlock, add /proc stub

October 12, 2007
	9p: add -n to disable authentication
	9pserve: add -n to reject authentication
	srv: add -n to reject authentication

October 9, 2007
	mount: ask for Unix test explicitly (Chris King)

September 24-25, 2007
	venti: many fixes, new index cache

September 16, 2007
	factotum: fix build error (Axel Belinfante)

September 15, 2007
	auxstats: use acpi for battery on Linux (Andrey Mirtchovski)

September 13, 2007
	factotum: implement 9pcr (of dubious utility)

August 28, 2007
	9pserve: fix infinite loop bug

August 23, 2007
	venti: add -s flag, bug fixes (Michael Kaminsky)

August 22, 2007
	9.rc: change to work with rc -e (Tony Lainson)
	9pserve: fix races causing assert(c->nmsg == 0) to fail.
	acme: fix warning memory leak (Michael Teichgräber)
	libfmt: add fmtlocale to Makefile (Faried Nawaz)
	libthread: simplify alt tracking, possibly fixing bug
	mailfs: fix handling of FLAGS responses (Robert Vollmert)
	unix ports: Linux x86-64 Makefiles (Ævar Arnfjörð Bjarmason)

July 25, 2007
	9pfuse: even more MacFUSE path changes

June 29, 2007
	dict: look in $dictpath for dictionaries

June 24, 2007
	mailfs: fix subject line in header file.

June 23, 2007
	mailfs: various IMAP bug fixes
	Mail: add Get command, refresh at startup,
		fix deleted messages bug
	nedmail: refresh at startup, on i and y.

June 20, 2007
	9pfuse: add -a aname to get at sourcesdump

June 19, 2007
	9pfuse: the MacFUSE file locations changed again (Rob Pike)

June 18, 2007
	upas/marshal: remove debugging print
	9pserve: ignore window resizes (fix really should be elsewhere)
	libdiskfs: more HFS fixes (David Swasey)

June 15, 2007
	acme(1): fix Snarf bug, change Indent ON to apply to all windows.

June 12, 2007
	vcat: always write last block, to make disk correct size
	vbackup: more OS X fixes, add -M flag
	vbackup(8): document -M

June 9, 2007
	libregexp, sam, acme: day 1 bug in sam regexp code

June 8, 2007
	libdiskfs: add HFS resource fork support (David Swasey)

June 5, 2007
	ramfs: be less chatty

June 3, 2007
	vbackup: fix dotdot (David Swasey)

May 29, 2007
	libdiskfs: more HFS fixes (David Swasey)
	
May 28, 2007
	libdiskfs: missing Bterm, make ext2 quieter,
		add disk partitions, HFS code (David Swasey)

May 21, 2007
	libregexp: fix pointer comparison warning (Ramon de Vera)
	test: fix
	
May 20, 2007
	touch: print error if wstat fails
	libc: use futimes on Linux to implement dirfwstat

May 19, 2007
	test: import from Plan 9

May 18, 2007
	libregexp: fix bug about leading ^ in regexp on multiline text (Roger Peppe)

May 15, 2007
	libc: new media size code for _p9dir on FreeBSD (Lou Kamenov)
	acme(4): document tab width field in ctl file

May 10, 2007
	libc: more NaN fixes (rsc)
		use device names in _p9dir for FreeBSD (Lou Kamenov)

src/cmd/page/page.c
#include <u.h>
#include <libc.h>
#include <draw.h>
#include <thread.h>
#include <bio.h>
#include <cursor.h>
#include "page.h"

int resizing;
int mknewwindow;
int doabort;
int chatty;
int reverse = -1;
int goodps = 1;
int ppi = 100;
int teegs = 0;
int truetoboundingbox;
int textbits=4, gfxbits=4;
int stdinfd;
int truecolor;
int imagemode;
int notewatcher;
int notegp;
int fitwin;
char tempfile[40];

int
watcher(void *v, char *x)
{
	USED(v);
	if(strcmp(x, "die") != 0)
		postnote(PNGROUP, notegp, x);
	threadexitsall(0);
	return 0;
}

void
watcherproc(void *v)
{
	threadnotify(watcher, 1);
	for(;;)
		sleep(1000);
}

int
bell(void *u, char *x)
{
	if(x && strcmp(x, "hangup") == 0)
		threadexitsall(0);

	if(x && strstr(x, "die") == nil)
		fprint(2, "postnote %d: %s\n", getpid(), x);

	/* alarms come from the gs monitor */
	if(x && strstr(x, "alarm")){
		postnote(PNGROUP, getpid(), "die (gs error)");
		postnote(PNPROC, notewatcher, "die (gs error)");
	}

	/* function mentions u so that it's in the stack trace */
	if((u == nil || u != x) && doabort)
		abort();

/*	fprint(2, "exiting %d\n", getpid()); */
	wexits("note");
	return 0;
}

static int
afmt(Fmt *fmt)
{
	char *s;

	s = va_arg(fmt->args, char*);
	if(s == nil || s[0] == '\0')
		return fmtstrcpy(fmt, "");
	else
		return fmtprint(fmt, "%#q", s);
}

void
usage(void)
{
	fprint(2, "usage: page [-biRrwf] [-p ppi] file...\n");
	wexits("usage");
}

void
cleanup(void)
{
	remove(tempfile);
}

void
threadmain(int argc, char **argv)
{
	Document *doc;
	Biobuf *b;
	enum { Ninput = 16 };
	uchar buf[Ninput+1];
	int readstdin, haveppi;

	haveppi = 0;
	ARGBEGIN{
	/* "temporary" debugging options */
	case 'P':
		goodps = 0;
		break;
	case 'v':
		chatty++;
		break;
	case 'V':
		teegs++;
		break;
	case 'a':
		doabort++;
		break;
	case 'T':
		textbits = atoi(EARGF(usage()));
		gfxbits = atoi(EARGF(usage()));
		break;

	/* real options */
	case 'R':
		resizing = 1;
		break;
	case 'r':
		reverse = 1;
		break;
	case 'p':
		haveppi = 1;
		ppi = atoi(EARGF(usage()));
		break;
	case 'b':
		truetoboundingbox = 1;
		break;
	case 'w':
		fprint(2, "warning: page -w only supported on x11 systems\n");
		resizing = 1;
		break;
	case 'i':
		imagemode = 1;
		break;
	case 'W':
		winsize = EARGF(usage());
		break;
	case 'f':
		fitwin = 1;
		break;
	default:
		usage();
	}ARGEND;

	notegp = getpid();

	notewatcher = proccreate(watcherproc, NULL, 1024);
	if(notewatcher == -1){
		sysfatal("proccreate");
		threadexitsall(0);
	}

	rfork(RFNOTEG);
	threadnotify(bell, 1);

	readstdin = 0;
	if(imagemode == 0 && argc == 0){
		readstdin = 1;
		stdinfd = dup(0, -1);
		close(0);
		open("/dev/tty", OREAD);
	}

	quotefmtinstall();
	fmtinstall('a', afmt);

	fmtinstall('R', Rfmt);
	fmtinstall('P', Pfmt);
	/*
	if(mknewwindow)
		newwin(); */

	if(readstdin){
		b = nil;
		if(readn(stdinfd, buf, Ninput) != Ninput){
			fprint(2, "page: short read reading %s\n", argv[0]);
			wexits("read");
		}

		atexit(cleanup);
	}else if(argc != 0){
		if(!(b = Bopen(argv[0], OREAD))) {
			fprint(2, "page: cannot open \"%s\"\n", argv[0]);
			wexits("open");
		}

		if(Bread(b, buf, Ninput) != Ninput) {
			fprint(2, "page: short read reading %s\n", argv[0]);
			wexits("read");
		}
	}else
		b = nil;

	if(initdraw(0, 0, "page") < 0){
		fprint(2, "page: initdraw failed: %r\n");
		wexits("initdraw");
	}
	display->locking = 1;
	ppi = scalesize(display, ppi);

	buf[Ninput] = '\0';
	if(imagemode)
		doc = initgfx(nil, 0, nil, nil, 0);
	else if(strncmp((char*)buf, "%PDF-", 5) == 0)
		doc = initpdf(b, argc, argv, buf, Ninput);
	else if(strncmp((char*)buf, "\x04%!", 2) == 0)
		doc = initps(b, argc, argv, buf, Ninput);
	else if(buf[0] == '\x1B' && strstr((char*)buf, "@PJL"))
		doc = initps(b, argc, argv, buf, Ninput);
	else if(strncmp((char*)buf, "%!", 2) == 0)
		doc = initps(b, argc, argv, buf, Ninput);
	else if(strcmp((char*)buf, "\xF7\x02\x01\x83\x92\xC0\x1C;") == 0)
		doc = initdvi(b, argc, argv, buf, Ninput);
	else if(strncmp((char*)buf, "\xD0\xCF\x11\xE0\xA1\xB1\x1A\xE1", 8) == 0)
		doc = initmsdoc(b, argc, argv, buf, Ninput);
	else if(strncmp((char*)buf, "x T ", 4) == 0)
		doc = inittroff(b, argc, argv, buf, Ninput);
	else {
		if(haveppi) {
			fprint(2, "page: you can't specify -p with graphic files\n");
			wexits("-p and graphics");
		}
		doc = initgfx(b, argc, argv, buf, Ninput);
	}

	if(doc == nil) {
		fprint(2, "page: error reading file: %r\n");
		wexits("document init");
	}

	if(doc->npage < 1 && !imagemode) {
		fprint(2, "page: no pages found?\n");
		wexits("pagecount");
	}

	if(reverse == -1) /* neither cmdline nor ps reader set it */
		reverse = 0;

	truecolor = screen->depth > 8;
	viewer(doc);
	wexits(0);
}

void
wexits(char *s)
{
	if(s && *s && strcmp(s, "note") != 0 && mknewwindow)
		sleep(10*1000);
	postnote(PNPROC, notewatcher, "die");
	postnote(PNGROUP, getpid(), "die");
	threadexitsall(s);
}
